packages/gatsby/src/utils/require-gatsby-plugin.ts

```diff
@@ -0,0 +1,24 @@
+const pluginModuleCache = new Map<string, any>()
+
+export function setGatsbyPluginCache(
+  plugin: { name: string; resolve: string },
+  module: string,
+  moduleObject: any
+): void {
+  const key = `${plugin.name}/${module}`
+  pluginModuleCache.set(key, moduleObject)
+}
+
+export function requireGatsbyPlugin(
+  plugin: { name: string; resolve: string },
+  module: string
+): any {
+  const key = `${plugin.name}/${module}`
+
+  let pluginModule = pluginModuleCache.get(key)
+  if (!pluginModule) {
+    pluginModule = require(`${plugin.resolve}/${module}`)
```

Comment:
This allows us to use node resolutions and exports inside package.json
```suggestion
    pluginModule = require(require.resolve(`${plugin.resolve}/${module}`))
```

Reply:
The suggested change fails with Jest. E.g. when trying to resolve `gatsby-plugin-test/gatsby-worker` (which is actually `gatsby-plugin-test/gatsby-worker.js`) - Jest resolver mock throws [here](https://github.com/facebook/jest/blob/3c511fd1a0a2e2d4c7a3c6880ac48e23cc807fa9/packages/jest-resolve/src/resolver.ts#L238-L241)

But it works with a regular `require`. Not sure if it's a bug with Jest or what. I've reverted this change, we can get back to it in a follow-up PR.

Reply:
Related: https://github.com/facebook/jest/issues/9543

Reply:
We are hitting this issue https://github.com/facebook/jest/issues/9543 - where jest (virtual) mocks have no impact on `require.resolve` 

There are workarounds (as done in https://github.com/gatsbyjs/gatsby/pull/28364 ) by using additional wrapper around `require.resolve` and then mocking that wrapper in jest and just "passtrough" id from argument.

But given we didn't use `require.resolve` up to this point, I would say that adding it is out of scope for this PR